qutip/propagator.py

```diff
@@ -236,8 +237,12 @@ def propagator(H, t, c_op_list=[], args={}, options=None,
             for n in range(N * N):
                 progress_bar.update(n)
                 col_idx, row_idx = np.unravel_index(n, (N, N))
-                rho0 = Qobj(sp.csr_matrix(([1], ([row_idx], [col_idx])),
-                                          shape=(N, N), dtype=complex))
+                rho0 = Qobj(
+                    sp.csr_matrix(
+                        ([1], ([row_idx], [col_idx])),
+                        shape=(N, N), dtype=complex),
+                    dims=H0.dims,
+                )
                 output = mesolve(H, rho0, tlist, c_op_list, [], args, options,
                                  _safe_mode=False)
                 for k, t in enumerate(tlist):
```

Comment:
This is a confusing way of writing `rho0 = projection(N, col_idx, row_idx); rho0.dims = H0.dims` (possibly with `col_idx` and `row_idx` the other way around).  Also, the `np.unravel_index` is quite confusing - it may be better just as `for row in range(N): for col in range(N)`.

Reply:
It would also be more efficient if we create a single `QobjEvo` out of `H` and `c_op_list` before this loop, and pass that as the first argument to `mesolve`.  (Similarly to before, I'd also not pass `e_ops` at all because it's empty, and pass the others by keyword.)

Reply:
I've made the `projection` and removal of `e_ops` changes because those make the code much more understandable, but I've left out the `QobjEvo` idea for now.

Reply:
Sure - definitely best to draw the line somewhere, or we'll end up doing an entire re-write bit-by-bit.